altair/vegalite/v5/api.py

```diff
@@ -732,15 +732,27 @@ def _parse_when(
 
 
 def _parse_literal(val: Any, *, str_as_lit: bool) -> dict[str, Any]:
-    if isinstance(val, str) and not str_as_lit:
-        return utils.parse_shorthand(val)
+    if isinstance(val, str):
+        return _str_as(val, "value" if str_as_lit else "field")
     elif _is_one_or_seq_literal_value(val):
         return {"value": val}
     else:
         msg = f"Expected one or more literal values, but got: {type(val).__name__!r}"
         raise TypeError(msg)
 
 
+def _str_as(val: str, to: Literal["datum", "field", "value"], /):
+    if to == "value":
+        return value(val)
+    elif to == "field":
+        return utils.parse_shorthand(val)
+    elif to == "datum":
+        return _ConditionClosed(test=_expr_core.GetAttrExpression(to, val))
```

Comment:
Not accessible yet. 
Purely guessing that this would be the correct shape. 

The returned type **must** support `.update()` and `**`, but cannot call `to_dict()` as `Expression` overrides it to return an `str` https://github.com/vega/altair/blob/c0f85cd1d1676f5fcdfda337d4990e24e93a9c39/altair/expr/core.py#L175-L176

Reply:
After searching for `alt.condition` in `r"./altair/tests/examples_arguments_syntax/*.py"`, I've found no examples of `datum` being used in this way. 

# Examples
All of these use some form of datum-based expression as the **predicate** only.

<details>
<summary>7 Examples</summary>


https://github.com/vega/altair/blob/c0f85cd1d1676f5fcdfda337d4990e24e93a9c39/tests/examples_arguments_syntax/bar_chart_with_highlighted_bar.py#L16-L20

https://github.com/vega/altair/blob/c0f85cd1d1676f5fcdfda337d4990e24e93a9c39/tests/examples_arguments_syntax/bar_chart_with_negatives.py#L15-L19

https://github.com/vega/altair/blob/c0f85cd1d1676f5fcdfda337d4990e24e93a9c39/tests/examples_arguments_syntax/candlestick_chart.py#L15-L17

https://github.com/vega/altair/blob/c0f85cd1d1676f5fcdfda337d4990e24e93a9c39/tests/examples_arguments_syntax/lasagna_plot.py#L11-L15

https://github.com/vega/altair/blob/c0f85cd1d1676f5fcdfda337d4990e24e93a9c39/tests/examples_arguments_syntax/layered_heatmap_text.py#L34-L38

https://github.com/vega/altair/blob/c0f85cd1d1676f5fcdfda337d4990e24e93a9c39/tests/examples_arguments_syntax/slider_cutoff.py#L24-L27

https://github.com/vega/altair/blob/c0f85cd1d1676f5fcdfda337d4990e24e93a9c39/tests/examples_arguments_syntax/us_employment.py#L31-L35

</details>

Comparing to where the `str_as`/`str_as_lit` argument in `alt.when` applies to `alt.condition`:

```py
import altair as alt

alt.condition(predicate, "maybe_str_1", "maybe_str_2")

alt.when(predicate).then("maybe_str_1").otherwise("maybe_str_2")
```
>  I think for altair it makes more sense if there was not str_as_lit, but the following items:
> - `str_as_value`
> - `str_as_datum`
> - `str_as_field`

> Edit: maybe easier to have str_as and then a Literal with the allowed options ['value', 'datum', 'field', ...]

@mattijn is there something I'm missing here?

If I am understanding this correctly, I think there is only a need for `str_as: Literal["shorthand", "value"]`.
The more specific options are covered already by default in [alt.utils.core.parse_shorthand](https://github.com/vega/altair/blob/c0f85cd1d1676f5fcdfda337d4990e24e93a9c39/altair/utils/core.py#L458-L650).

If this were a performance concern, I would focus on `parse_shorthand` first - with changes like this earlier commit https://github.com/vega/altair/commit/14f6ea162b50a1cd2cf0c9678deb46f96c03b1a1

Reply:
> If this were a performance concern, I would focus on `parse_shorthand` first ...

I was curious, but even with the changes below I wasn't able to observe any performance improvement when running `hatch test --all`:

<details>
<summary>Code block <b>altair/altair/utils/core.py</b></summary>

```py
# altair/altair/utils/core.py

...

_InferOption = Literal["aggregates", "window_ops", "timeunits", "types"]
_InferConfig: TypeAlias = Tuple[Tuple[_InferOption, bool], ...]

_SHORTHAND_PATTERN_CACHE: dict[_InferConfig, list[re.Pattern]] = {}

# I also tried adding a 2nd layer of caching here with `@functools.lru_cache(None)`
# Still saw no performance benefit
def _find_matches(
    shorthand: str, *, aggregates: bool, window_ops: bool, timeunits: bool, types: bool
) -> dict[str, str | Any]:
    # NOTE:
    # - Result of regexps is the cache target
    # - keyed to the `parse_` args
    # - `TopLevelMixin.transform_[timeunit|window]`
    #  - are only library code that uses non-default
    global _SHORTHAND_PATTERN_CACHE
    config: _InferConfig = (
        ("aggregates", aggregates),
        ("window_ops", window_ops),
        ("timeunits", timeunits),
        ("types", types),
    )

    def _generate() -> list[re.Pattern]:
        fmts: list[str] = []
        if aggregates:
            fmts.extend([r"{agg_count}\(\)", r"{aggregate}\({field}\)"])
        if window_ops:
            fmts.extend([r"{op_count}\(\)", r"{window_op}\({field}\)"])
        if timeunits:
            fmts.append(r"{timeUnit}\({field}\)")
        fmts.append(r"{field}")

        it = chain.from_iterable((p + ":{type}", p) for p in fmts) if types else fmts
        patterns = [
            re.compile(rf"\A{p.format(**SHORTHAND_UNITS)}\Z", re.DOTALL) for p in it
        ]
        _SHORTHAND_PATTERN_CACHE[config] = patterns
        return patterns

    patterns = _SHORTHAND_PATTERN_CACHE.get(config, _generate())
    return next(m.groupdict() for pat in patterns if (m := pat.match(shorthand)))


def parse_shorthand(
    shorthand: dict[str, Any] | str,
    data: pd.DataFrame | DataFrameLike | None = None,
    parse_aggregates: bool = True,
    parse_window_ops: bool = False,
    parse_timeunits: bool = True,
    parse_types: bool = True,
) -> dict[str, Any]:
    """General tool to parse shorthand values ..."""
    from altair.utils._importers import pyarrow_available

    if not shorthand:
        return {}
    # find matches depending on valid fields passed
    elif isinstance(shorthand, dict):
        # NOTE: This branch previously did all the same initial work,
        # but then used `shorthand` every time
        attrs = shorthand
    else:
        attrs = _find_matches(
            shorthand,
            aggregates=parse_aggregates,
            window_ops=parse_window_ops,
            timeunits=parse_timeunits,
            types=parse_types,
        )

    # Handle short form of the type expression
    if "type" in attrs:
        attrs["type"] = INV_TYPECODE_MAP.get(attrs["type"], attrs["type"])

    # counts are quantitative by default
    if attrs == {"aggregate": "count"}:
        attrs["type"] = "quantitative"

    # times are temporal by default
    if "timeUnit" in attrs and "type" not in attrs:
        attrs["type"] = "temporal"

    # if data is specified and type is not, infer type from data
    if "type" not in attrs:
        if pyarrow_available() and data is not None and isinstance(data, DataFrameLike):
            dfi = data.__dataframe__()
            if "field" in attrs:
                unescaped_field = attrs["field"].replace("\\", "")
                if unescaped_field in dfi.column_names():
                    column = dfi.get_column_by_name(unescaped_field)
                    try:
                        attrs["type"] = infer_vegalite_type_for_dfi_column(column)
                    except (NotImplementedError, AttributeError, ValueError):
                        # Fall back to pandas-based inference.
                        # Note: The AttributeError catch is a workaround for
                        # https://github.com/pandas-dev/pandas/issues/55332
                        if isinstance(data, pd.DataFrame):
                            attrs["type"] = infer_vegalite_type(data[unescaped_field])
                        else:
                            raise

                    if isinstance(attrs["type"], tuple):
                        attrs["sort"] = attrs["type"][1]
                        attrs["type"] = attrs["type"][0]
        elif isinstance(data, pd.DataFrame):
            # Fallback if pyarrow is not installed or if pandas is older than 1.5
            #
            # Remove escape sequences so that types can be inferred for columns with special characters
            if "field" in attrs and attrs["field"].replace("\\", "") in data.columns:
                attrs["type"] = infer_vegalite_type(
                    data[attrs["field"].replace("\\", "")]
                )
                # ordered categorical dataframe columns return the type and sort order as a tuple
                if isinstance(attrs["type"], tuple):
                    attrs["sort"] = attrs["type"][1]
                    attrs["type"] = attrs["type"][0]

    # If an unescaped colon is still present, it's often due to an incorrect data type specification
    # but could also be due to using a column name with ":" in it.
    if (
        "field" in attrs
        and ":" in attrs["field"]
        and attrs["field"][attrs["field"].rfind(":") - 1] != "\\"
    ):
        raise ValueError(
            '"{}" '.format(attrs["field"].split(":")[-1])
            + "is not one of the valid encoding data types: {}.".format(
                ", ".join(TYPECODE_MAP.values())
            )
            + "\nFor more details, see https://altair-viz.github.io/user_guide/encodings/index.html#encoding-data-types. "
            + "If you are trying to use a column name that contains a colon, "
            + 'prefix it with a backslash; for example "column\\:name" instead of "column:name".'
        )
    return attrs

...

```

</details>

Thought I'd leave this here, since I won't be committing but it may be helpful to know this part isn't a performance bottleneck.